services/src/main/java/org/keycloak/services/resources/account/LinkedAccountsResource.java

```diff
@@ -0,0 +1,208 @@
+ * limitations under the License.
+ */
+package org.keycloak.services.resources.account;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.UUID;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriBuilder;
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.keycloak.common.util.Base64Url;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.events.Details;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.AccountRoles;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.FederatedIdentityModel;
+import org.keycloak.models.IdentityProviderModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.services.ErrorResponse;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.Auth;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.services.validation.Validation;
+
+/**
+ * API for linking/unlinking social login accounts
+ *
+ * @author Stan Silvert
+ */
+public class LinkedAccountsResource {
+    private static final Logger logger = Logger.getLogger(LinkedAccountsResource.class);
+    
+    private final KeycloakSession session;
+    private final HttpRequest request;
+    private final ClientModel client;
+    private final EventBuilder event;
+    private final UserModel user;
+    private final RealmModel realm;
+    private final Auth auth;
+
+    public LinkedAccountsResource(KeycloakSession session, 
+                                  HttpRequest request, 
+                                  ClientModel client,
+                                  Auth auth, 
+                                  EventBuilder event, 
+                                  UserModel user) {
+        this.session = session;
+        this.request = request;
+        this.client = client;
+        this.auth = auth;
+        this.event = event;
+        this.user = user;
+        realm = session.getContext().getRealm();
+    }
+    
+    @GET
+    @Path("/")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response linkedAccounts() {
+        LinkedAccountsBean linkedAccounts = new LinkedAccountsBean(this.session, this.realm, this.user);
+        return Cors.add(request, Response.ok(linkedAccounts)).auth().allowedOrigins(auth.getToken()).build();
+    }
+    
+    @GET
+    @Path("/{providerId}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response buildLinkedAccountURI(@PathParam("providerId") String providerId, 
```

Comment:
The URL should be built on the client side, not on the server side. So this whole method can be removed.

Reply:
I like the rest of your suggestions, but disagree on this one.  Generating this URL is rather complicated.  If we do it on the client side then every application that wants to build the URL needs to write code for it.  Also, if the requirements for the URL change then every client needs to rewrite and re-test their code.

With this convenience API, the code is only written once and it is maintained in one place.  

All that being said, it brings up the question of why the broker API is so complicated in the first place?  Why can't you just call the broker with the provider ID and redirect URL?

Reply:
The reason the linking API is complicated has to do with security issues. For this to work it has to be invoked as it is from the client and it's the clients responsibility to generate these links as well as verify on the response. Its basically down to guarding against vulnerabilities in a redirect based API.

By introducing this you are effectively bypassing the protection here. We should definitively not introduce an alternative and less secure API.

Reply:
I don't see how it's less secure.  It ends up being exactly the same thing as we have today with the URL generated on the server.

Reply:
The checks are there for a reason and they should be done on the client side, not the server side. With the current FreeMarker based account management console they are being done on the client side, it just happens that it's a server-side app that runs in Keycloak itself.

The nonce/hash are there for an application to know that it initiated the request when it receives the redirect back. It's for the initiating client to verify. With these being generated on the server-side there is no way for the initiating client to verify these as it doesn't have access to them.

It really does have to be this way. If it didn't the linking API wouldn't have been designed in the way it is.

Reply:
To explain further, today's account management sends a request to the server.  The server generates the url and sends it back to the client as a redirect to the broker.

What I am doing is the same.  I send a request to the server and it sends back the URL.  The only difference is that the URL is sent back as JSON instead of as a redirect.  Again, this is a convenience API that makes it easier for my Angular app and easier for an app written in another language or framework.  From a security standpoint, it's the same as today's account management console.

There is no security concern to just build the initial url, which is all that my code is doing.   The broker is the part that is worried about security.  My API is not dealing with the broker.  It's just creating a URL for you to use.

Reply:
I still think you are mistaken.  The verification you are talking about happens inside the broker, not inside the client.

You can look at what happens today:
Client creates URL with nonce and sends to broker (with nonce)
Broker redirects to social provider (with nonce)
Social provider redirects back to broker (with nonce)
Broker redirects back to account management (**without** nonce)

It's no different from what I do in this PR.